composer/distributed/fsdp2.py

```diff
@@ -140,3 +141,27 @@ def prepare_fully_shard(
             if attr == 'verbose':
                 continue
             log.info(f'FSDP2: {attr}: {getattr(fsdp2_config, attr)}')
+
+
+def sync_module_states(model: nn.Module, full_state_dict: dict) -> None:
+    """Syncs the module states of the model across all ranks using the full state dict.
+
+    This function synchronizes the module states of the model across all ranks using the full state dict.
+    This handles mixed initialization scenarios where different ranks have parameters on different devices.
+
+    Args:
+        model (nn.Module): The FSDP2-wrapped model to synchronize.
+        full_state_dict (dict): The full state dict to synchronize. This is only fully populated on rank 0. The other ranks will receive a partial state dict.
+
+    Returns:
+        None
+    """
+    from torch.distributed.checkpoint.state_dict import StateDictOptions, set_model_state_dict
+
+    if dist.get_global_rank() == 0:
+        model = model.to(device=torch.cuda.current_device(), non_blocking=True)
+    else:
+        model = model.to_empty(device=torch.cuda.current_device())
+
+    options = StateDictOptions(full_state_dict=True, broadcast_from_rank0=True, strict=False)
+    set_model_state_dict(model, full_state_dict, options=options)
```

Comment:
we probably need to broadcast buffers here too?

Reply:
the buffers would exist as part of the `state_dict`, right? I guess https://github.com/volcengine/verl/blob/main/verl/utils/fsdp_utils.py#L416-L418 mentions that there are some buffers that aren't part of the `state_dict` that would be best to broadcast as well though so I will add it.